feat: let_projs tactic, recursively adding projections of hypotheses

semorrison opened the pull request:
`let_projs` adds let bindings for all projections of local hypotheses, recursively.

It's probably rarely useful interactively, but may be useful as a preprocessing step for tactics such as `solve_by_elim`.

digama0 commented:
I'm inclined to not surface this as a user-facing tactic at all, and just have a `MetaM` tactic for it. This introduces more API surface area which we have to avoid breaking later, without a compelling use case.

semorrison commented:
Good idea. I mostly wrote the syntactical frontend so I could test it. I've moved the syntax into the test file, and added a note that it is not useful otherwise.

nomeata commented on Std/Tactic/LetProjs.lean:
```diff
@@ -0,0 +1,82 @@
+def allProjs (e : Expr) : MetaM (Array (Name × Expr)) := do
+  let (c, _) := (← inferType e).getAppFnArgs
+  let env ← getEnv
+  unless isStructure env c do return #[]
+  (getStructureFields! env c).filterMapM fun f => return (f, ← mkProjection e f)
```
`getStructureFields!" already returns `#[]` if it’s not a structure, so maybe
```suggestion
  (getStructureFields! env c).filterMapM fun f => return (f, ← mkProjection e f)
```
and maybe don’t name it with `!`?

semorrison replied:
I've just inlined it.

nomeata commented:
Is there no worry that eagerly adding all projections is going to blow up in some cases, with large or deeply nested structures? Or is this only going to be used in places where the user expects to wait a bit longer (interactive `exact?` maybe)? Or is this simply a normal approach?

semorrison commented:
> Is there no worry that eagerly adding all projections is going to blow up in some cases, with large or deeply nested structures? Or is this only going to be used in places where the user expects to wait a bit longer (interactive `exact?` maybe)? Or is this simply a normal approach?

@joehendrix's suggestion is that when other tactics are using this, if it is a user-called tactic it's okay to do something potentially expensive like this, but tactics called by other tactics should not.
